src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Hover/RazorHoverEndpoint.cs

```diff
@@ -5,88 +5,106 @@
             DocumentContextFactory documentContextFactory,
             RazorHoverInfoService hoverInfoService,
+            LanguageServerFeatureOptions languageServerFeatureOptions,
+            RazorDocumentMappingService documentMappingService,
+            ClientNotifierServiceBase languageServer,
             ILoggerFactory loggerFactory)
+            : base(documentContextFactory, languageServerFeatureOptions, languageServer, loggerFactory.CreateLogger<RazorHoverEndpoint>())
         {
-            if (documentContextFactory is null)
-            {
-                throw new ArgumentNullException(nameof(documentContextFactory));
-            }
+            _documentContextFactory = documentContextFactory ?? throw new ArgumentNullException(nameof(documentContextFactory));
+            _hoverInfoService = hoverInfoService ?? throw new ArgumentNullException(nameof(hoverInfoService));
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+        }
 
-            if (hoverInfoService is null)
-            {
-                throw new ArgumentNullException(nameof(hoverInfoService));
-            }
+        public RegistrationExtensionResult? GetRegistration(VSInternalClientCapabilities clientCapabilities)
+        {
+            const string AssociatedServerCapability = "hoverProvider";
+            _clientCapabilities = clientCapabilities;
 
-            if (loggerFactory is null)
+            var registrationOptions = new HoverOptions()
             {
-                throw new ArgumentNullException(nameof(loggerFactory));
-            }
+                WorkDoneProgress = false,
+            };
 
-            _documentContextFactory = documentContextFactory;
-            _hoverInfoService = hoverInfoService;
-            _logger = loggerFactory.CreateLogger<RazorHoverEndpoint>();
+            return new RegistrationExtensionResult(AssociatedServerCapability, registrationOptions);
         }
 
-        public async Task<VSInternalHover?> Handle(VSHoverParamsBridge request, CancellationToken cancellationToken)
+        /// <inheritdoc/>
+        protected override string CustomMessageTarget => RazorLanguageServerCustomMessageTargets.RazorHoverEndpointName;
+
+        /// <inheritdoc/>
+        protected override async Task<DelegatedHoverParams> CreateDelegatedParamsAsync(VSHoverParamsBridge request, DocumentContext documentContext, CancellationToken cancellationToken)
         {
-            if (request is null)
-            {
-                throw new ArgumentNullException(nameof(request));
-            }
+            var sourceText = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+            var absoluteIndex = request.Position.GetRequiredAbsoluteIndex(sourceText, Logger);
+            var projection = await _documentMappingService.GetProjectionAsync(documentContext, absoluteIndex, cancellationToken).ConfigureAwait(false);
```

Comment:
Interesting thought, do you feel like we'll always be getting the projection twice per-handler? This bit isn't cached and is re-computed which makes me think we should either cache this in the underlying mapping service OR we should bake projection creation into the abstract base type.

Reply:
I'm not sure yet. I know we'll always be getting the context, but projection _may_ not be needed. We can optimize a bit more as we add more handlers I think